### PR TITLE
#213 - Validation: use default language if no language is given in options

### LINKED ISSUE #213: Validation: use default language if not provided
## Proposal

When using validation manually, not providing the `Language` field in the `validation.Options` will trigger a `panic: runtime error: invalid memory address or nil pointer dereference`.

We should load the built-in default language (`en-US`) if the option is omitted. Or we could use an empty language that would return only the name of the entry.

### Possible drawbacks

Loading the default language for every validation can reduce performance a bit. This would still be very much recommended to use the option. This would need to be properly documented.

### PR DESCRIPTION
**Pull request by BowlOfSoup**
Fetch default language if no language is given when validating.

## References

**Issue:** fixes #213

### CODE REVIEW

File: validation/validator_test.go [hunk 1]
```diff
@@ -282,6 +282,23 @@ func TestValidate(t *testing.T) {
 			},
 			wantData: map[string]any{},
 		},
+		{
+			desc: "nil_delete_from_parent_without_language",
```
Comment (System-Glitch) [inline validation/validator_test.go:286]:
```suggestion
			desc: "no_language_provied_in_options",
```

File: validation/validator.go [hunk 2]
```diff
@@ -275,6 +275,9 @@ func Validate(options *Options) (*Errors, []error) {
 	if options.Extra == nil {
 		options.Extra = map[any]any{}
 	}
+	if options.Language == nil {
+		options.Language = lang.New().GetDefault()
```
Comment (System-Glitch) [inline validation/validator.go:279]: Please add a comment on the `Language` field in the `Options` structure:

```
// Language used for validation error message translation. If `nil` or omitted, defaults to `en-US`.
```

**Review by System-Glitch — APPROVED** [review]
LGTM 👍🏻

**Comment by BowlOfSoup** [comment]
> @BowlOfSoup Is this PR ready?

@System-Glitch no it's not. And I even think now we shouldn't fix this at all:
1. Performance penalty
2. Too easy to abuse the possibility of not passing a language. People may not care.

It's up to you to decline 🙂

**Comment by System-Glitch** [comment]
I will think about it. Thank you for your input.

**Comment by System-Glitch** [comment]
To solve the performance issue, let's go back to using an empty language and return the lang entry name:
```go
if options.Language == nil {
	options.Language = &lang.Language{}
}
```
This will require a slight change in the `lang/language.go` file to make sure a `*Language` doesn't panic if its maps are not initialized.

I think it's a good balance between performance and convenience. We'll just need to properly document what happens if the option is not provided.

**Comment by BowlOfSoup** [comment]
> To solve the performance issue, let's go back to using an empty language and return the lang entry name:
> 
> ```go
> if options.Language == nil {
> 	options.Language = &lang.Language{}
> }
> ```
> 
> This will require a slight change in the `lang/language.go` file to make sure a `*Language` doesn't panic if its maps are not initialized.
> 
> I think it's a good balance between performance and convenience. We'll just need to properly document what happens if the option is not provided.

This would return untranslated strings. Is this the possibility we want to give the user? We could, then input = output.

**Comment by System-Glitch** [comment]
> This would return untranslated strings. Is this the possibility we want to give the user? We could, then input = output.
I prefer this over a nil dereference panic.

Another solution would be to return an error (or panic) with an explicit message saying that this option is mandatory.

I think it would be nice to have someone else's opinion on this.

**Comment by System-Glitch** [comment]
@BowlOfSoup I had an idea that would satisfy both the need for translated lines as well as performance:
- Add a getter for the default language in `lang/default.go`:
```go
func GetDefault() *Language {
	return enUS
}
```
- Use this in `validation/validator.go` if the language option is not provided.

The downside is that custom language entries won't be supported this way since only the builtin default ones are loaded. I think this is acceptable. We just need to document it right to make it clear.

**Comment by BowlOfSoup** [comment]
This would indeed be a good solution!Op 26 jul 2024 om 10:30 heeft SystemGlitch ***@***.***> het volgende geschreven:﻿
@BowlOfSoup I had an idea that would satisfy both the need for translated lines as well as performance:

Add a getter for the default language in lang/default.go:

func GetDefault() *Language {
	return enUS
}

Use this in validation/validator.go if the language option is not provided.

The downside is that custom language entries won't be supported this way since only the builtin default ones are loaded. I think this is acceptable. We just need to document it right to make it clear.

**Comment by System-Glitch** [comment]
I implemented it in this commit: 3e4f090a2c5ac7c02cb17cf47a64d6958f9f5ffa

Thank you very much for your suggestion!

**Pull request closed without merging**